runtime/expr/values.go

```diff
@@ -275,48 +243,105 @@ type Entry struct {
-	b := m.builder
-	b.Reset()
+	m.builder.Reset()
+	m.keys.reset()
+	m.vals.reset()
 	for _, e := range m.entries {
-		key := e.Key.Eval(ectx, this)
-		val := e.Val.Eval(ectx, this)
-		if keyType == nil {
-			if m.typ == nil || m.typ.KeyType != key.Type || m.typ.ValType != val.Type {
-				keyType = key.Type
-				valType = val.Type
-				m.typ = m.zctx.LookupTypeMap(keyType, valType)
-			} else {
-				keyType = m.typ.KeyType
-				valType = m.typ.ValType
-			}
-		} else if keyType != m.typ.KeyType || valType != m.typ.ValType {
-			//XXX issue #3363
-			return ectx.CopyValue(*m.zctx.NewErrorf("mixed-type map expressions not yet supported"))
-		}
-		b.Append(key.Bytes)
-		b.Append(val.Bytes)
+		m.keys.append(e.Key.Eval(ectx, this))
+		m.vals.append(e.Val.Eval(ectx, this))
 	}
-	bytes := b.Bytes()
-	if bytes == nil {
-		// Return empty map instead of null map.
-		bytes = []byte{}
+	if len(m.keys.types) == 0 {
+		typ := m.zctx.LookupTypeMap(zed.TypeNull, zed.TypeNull)
+		return ectx.NewValue(typ, []byte{})
+	}
+	kIter, vIter := m.keys.iter(m.zctx), m.vals.iter(m.zctx)
+	for !kIter.done() {
+		kIter.appendNext(&m.builder)
+		vIter.appendNext(&m.builder)
+	}
+	bytes := m.builder.Bytes()
+	typ := m.zctx.LookupTypeMap(kIter.typ, vIter.typ)
+	return ectx.NewValue(typ, zed.NormalizeMap(bytes))
+}
+
+type collectionBuilder struct {
+	types       []zed.Type
+	uniqueTypes []zed.Type
+	bytes       []zcode.Bytes
+}
+
+func (c *collectionBuilder) reset() {
+	c.types = c.types[:0]
+	c.uniqueTypes = c.uniqueTypes[:0]
+	c.bytes = c.bytes[:0]
+}
+
+func (c *collectionBuilder) append(val *zed.Value) {
+	c.types = append(c.types, val.Type)
+	c.bytes = append(c.bytes, val.Bytes)
+}
+
+func (c *collectionBuilder) iter(zctx *zed.Context) collectionIter {
+	// uniqueTypes must be copied since zed.UniqueTypes operates on the type
+	// array in place and thus we'll lose order.
+	for _, typ := range c.types {
+		c.uniqueTypes = append(c.uniqueTypes, typ)
+	}
+	return collectionIter{
+		typ:   unionOf(zctx, c.uniqueTypes),
+		bytes: c.bytes,
+		types: c.types,
+	}
+}
+
+type collectionIter struct {
+	typ   zed.Type
+	bytes []zcode.Bytes
+	types []zed.Type
+}
+
+func (c *collectionIter) appendNext(b *zcode.Builder) {
+	if union, ok := c.typ.(*zed.TypeUnion); ok {
+		zed.BuildUnion(b, union.Selector(c.types[0]), c.bytes[0])
+	} else {
+		b.Append(c.bytes[0])
+	}
+	c.bytes = c.bytes[1:]
+	c.types = c.types[1:]
+}
+
+func (c *collectionIter) done() bool {
+	return len(c.types) == 0
+}
+
+func unionOf(zctx *zed.Context, types []zed.Type) zed.Type {
+	types = zed.UniqueTypes(types)
+	if len(types) == 1 {
+		return types[0]
+	}
+	if len(types) == 2 {
+		// If there's only two types and one type is null, return single
+		// non-null type.
+		if types[0] == zed.TypeNull {
+			return types[1]
+		} else if types[1] == zed.TypeNull {
+			return types[0]
+		}
 	}
-	return ectx.CopyValue(zed.Value{m.typ, zed.NormalizeMap(bytes)})
+	return zctx.LookupTypeUnion(types)
```

Comment:
I think we should discard the null type whenever we have at least one other type.
```go
	 unique := types[:0]
	 for _, t := range zed.UniqueTypes(types) {
		if t != zed.TypeNull {
			unique = append(unique, t)
		}
	}
	if len(unique) == 0 {
		return zed.TypeNull
	}
	if len(unique) == 1 {
		return types[0]
	}
	return zctx.LookupTypeUnion(types)
```

Reply:
What then would the selector of the null value be then?

Reply:
Discussed this offline, will fix.